### PR TITLE
Scaffolding: monorepo, server, types, linting, monitoring, ci/cd

### PR DESCRIPTION
**Pull request by promaty**
You will see some new things here:
- koa
This is a nice upgrade from Express, notice how little code I need to do async middleware
- typescript
This was complicated to configure (especially for monorepo), but I like how it turned out.
A common problem with typescript is that compilation is long. For this reason I used babel typescript plugin which is supported by typescript: https://devblogs.microsoft.com/typescript/typescript-and-babel-7/
TL;DR babel simply removes the types without type checking, so it's super fast and we can also run `nodemon` with `babel-node` for development. When ready to build I run the `tsc` command for type checking.
Also I've applied some recommended `eslint` rules.

Most of other parts are just copied from court-backend. Although I had to rewrite some modules specifically for Koa+Typescript.

### CODE REVIEW

**Comment by promaty** [comment]
Also forgot to mention:
- yarn workspaces
This allows having one `node_modules` at the top for all packages. Also I've added main dev dependencies at the top `package.json` so we don't have to install things like babel multiple times.

File: .gitignore [hunk 1]
```diff
@@ -0,0 +1,4 @@
+node_modules
+lib
```
Comment (facuspagnuolo) [inline .gitignore:2]: `lib` sounds really general, if it's being used for the typescript compilation, should we rename it to `build` or `dist`? honestly not sure what has been adopted by the community

File: packages/server/test/integration/file-creation.test.ts [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+  test('should create a file', async () => {
+    const res = await request(serverURL).post('/files').send({
+      created: true
+    })
+    expect(res.status).toEqual(HttpStatus.OK)
+    expect(res.body).toEqual({
+      created: true
+    })
+  })
+})
```
Comment (facuspagnuolo) [inline packages/server/test/integration/file-creation.test.ts:17]: Nice first test :)

File: .gitignore (hunk 1 above)
Comment (promaty) [reply .gitignore:2]: will rename to `build`

**Comment by promaty** [comment]
> Honestly, I was thinking you were not going to use lerna for this. What was the rationale? Do you see further sub-packages appearing here?

From the architecture comments it felt like we are going to end up using some background worker for delayed verification. I also solved some problems here that we have in `court-backend`, so I can fix that one easier now (using babel in `shared` for models, building before others)

File: packages/server/test/integration/file-creation.test.ts (hunk 2 above)
Comment (promaty) [reply packages/server/test/integration/file-creation.test.ts:17]: `jest` looks really good for testing, don't have to assemble multiple modules.

File: .github/workflows/dashboard.yml [hunk 3]
```diff
@@ -0,0 +1,20 @@
+name: Dashboard CI/CD
+on:
+  push:
+    branches:
+    - master
```
Comment (sohkai) [inline .github/workflows/dashboard.yml:5]: More of a stylistic discussion, but for yaml files, how do you feel about adding an indent when we go into list mode?

E.g.

```yaml
branches:
  - master
```

I see our k8s configs are done with them in the same line, but I was thinking it'd be nice to standardize one on or the other as having both can be confusing.

File: .github/workflows/dashboard.yml [hunk 4]
```diff
@@ -0,0 +1,20 @@
+name: Dashboard CI/CD
+on:
+  push:
+    branches:
```
Comment (sohkai) [inline .github/workflows/dashboard.yml:4]: I wonder if we should only run this with a release? It seems a bit weird that the dashboard would update ahead of the actual deployment :).

File: README.md [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+
+This is a mono-repo with the following packages:
+
+- [`server`](./packages/server): Exposes the REST API.
+- [`shared`](./packages/shared): Packages shared among other sub-repos.
+
+
+## Local Docker setup
+
+Development environment is configured using [docker-comopse](https://docs.docker.com/compose/).
```
Comment (sohkai) [inline README.md:13]: **docker-compose** 😄

File: packages/server/package.json [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+    "lint": "eslint src/** --ext .ts",
+    "test": "jest",
+    "build": "yarn lint && yarn check-types && yarn build:js",
+    "build:js": "babel ./src --out-dir ./build --extensions .ts --root-mode upward",
+    "build:shared": "lerna run build --scope */*-shared --stream",
+    "build:clean": "yarn clean && yarn build",
+    "clean": "rm -rf ./build",
+    "check-types": "tsc",
+    "start": "node ./build/index.js",
+    "start:dev": "nodemon --exec 'babel-node --extensions .ts --root-mode upward' ./src/index.ts"
```
Comment (sohkai) [inline packages/server/package.json:15]: Not sure with yarn workspaces, but it seems a bit weird to not declare the dev dependencies needed for the package in its own package.json (e.g. here, we don't declare `babel`, `eslint`, `nodemon`, `tsc`, etc).

File: .eslintrc.json [hunk 7]
```diff
@@ -0,0 +1,9 @@
+{
+  "root": true,
+  "plugins": ["@typescript-eslint"],
+  "parser": "@typescript-eslint/parser",
```
Comment (andy-hook) [inline .eslintrc.json:4]: Worth checking but you _might_ need to provide a [project](https://www.npmjs.com/package/@typescript-eslint/parser#user-content-parseroptionsproject) with optional [tsconfigRootDir](https://www.npmjs.com/package/@typescript-eslint/parser#tsconfigRootDir) to have the linting match type errors reported at build.

**Comment by andy-hook** [comment]
>   A common problem with typescript is that compilation is long. For this reason I used babel typescript plugin which is supported by typescript

❤️

File: .eslintrc.json (hunk 7 above)
Comment (promaty) [reply .eslintrc.json:4]: Seems that this is only required for rules that require type information, I'm not using those for now (e.g. `plugin:@typescript-eslint/recommended-requiring-type-checking`)

File: monitoring/prometheus/prometheus.yml [hunk 8]
```diff
@@ -0,0 +1,10 @@
+# This config is required for Prometheus to scrape other containers
+scrape_configs:
+- job_name: ipfs-background-service
+  static_configs:
+  - targets: 
+    - server:9091
```
[the remark this answers is not in the document]
Comment (promaty) [reply monitoring/prometheus/prometheus.yml:6]: Yes I couldn't find a way to use .env here. I also have the ports hardcoded in k8s in some cases, but I try to use a port name/var whenever I can.

File: packages/server/package.json (hunk 6 above)
Comment (promaty) [reply packages/server/package.json:15]: It's the same dependency for every package so I didn't want to repeat myself. Also if we need to update typescript/babel  we have to make sure we do it in every package, I'm sure I will mess it up at some point. We always do development on the whole monorepo anyways so I'm not sure about the value of declaring this multiple times

File: .github/workflows/dashboard.yml (hunk 3 above)
Comment (sohkai) [reply .github/workflows/dashboard.yml:5]: OK—let's standardize on the non-nested version; you probably write the most yaml out of everyone :).

File: packages/server/package.json (hunk 6 above)
Comment (sohkai) [reply packages/server/package.json:15]: I usually feel more comfortable declaring all the dependencies with the package itself, but definitely agree there's a lot of leeway here since all the components of the monorepo are meant to be used together!

If we can, let's leave a note in the README about this—that all shared dev dependencies are declared at the top level.

File: .github/workflows/dashboard.yml (hunk 3 above)
Comment (promaty) [reply .github/workflows/dashboard.yml:5]: professional yaml developer
